spec/registry.md

```diff
@@ -384,6 +322,158 @@ its _resolved value_ contains the implementation-defined integer value
 of the _operand_ of the annotated _expression_,
 together with the resolved options' values.
 
+### The `:currency` function
+
+The function `:currency` is a selector and formatter for currency values, 
+which are a specialized form of numeric selection and formatting.
+
+#### Operands
+
+The _operand_ of the `:currency` function can be one of any number of
+implementation-defined types,
+each of which contains a numerical `value` and a `currency`;
+or it can be a [Number Operand](#number-operands), as long as the option
+`currency` is provided.
+The option `currency` MUST NOT be used to override the currency of an implementation-defined type.
+Using this option in such a case results in a _Bad Option_ error.
+
+The value of the _operand_'s `currency` MUST be either a string containing a
+valid [Unicode Currency Identifier](https://cldr-smoke.unicode.org/spec/main/ldml/tr35.html#UnicodeCurrencyIdentifier)
+or an implementation-defined currency type.
+
+A [Number Operand](#number-operands) without a `currency` _option_ results in a _Bad Operand_ error.
+
+> [!NOTE]
+> For example, in ICU4J, the type `com.ibm.icu.util.CurrencyAmount` can be used
+> to set the amount and currency.
+
+> [!NOTE]
+> For runtime environments that do not provide a ready-made data structure,
+> class, or type for currency values, the implementation ought to provide
+> a data structure, convenience function, or documentation on how to encode
+> the value and currency code for formatting.
+> For example, such an implementation might define a "currency operand"
+> to include a key-value structure with specific keys to be the
+> local currency operand, which might look like the following:
+> ```json
+> {
+>    "value": 123.45,
+>    "currency": "EUR"
+> }
+> ```
+
+#### Options
+
+Some options do not have default values defined in this specification.
+The defaults for these options are implementation-dependent.
+In general, the default values for such options depend on the locale, 
+the currency,
+the value of other options, or all of these.
+
+Fraction digits for currency values behave differently than for other numeric formatters.
+The number of fraction digits displayed is usually set by the currency used.
+For example, USD uses 2 fraction digits, while JPY uses none.
+Setting some other number of `fractionDigits` allows greater precision display
+(such as when performing currency conversions or other specialized operations)
+or disabling fraction digits if set to `0`.
+The special _option_ _value_ `hideIfWhole` is used to display values without
+fraction digits when the number of fraction digits is zero,
+or based on the currency when the number of fraction digits for the currency is non-zero.
```

Comment:
This should be left out. This is novel currency formatting behaviour, and we should not be introducing it here via message formatting. Is this already somehow supported in ICU4J and ICU4C? If this makes sense for formatting currencies, wouldn't this also make sense for many other number formatting situations?

I do not object to a value like this in principle, but I object to it being introduced _here_.

Reply:
This value is supported by ICU.

The use cases for currencies are more common than for other number values (with the possible exception of units). Normally one is trying to get consistent widths (or one doesn't care). With currency values you generally want a specific number of fractional digits or no fraction digits. Mutation of the value to produce this is undesirable.

However, I agree that this is "special sauce" and shouldn't be _required_ here. Should we introduce **optional** values?

Reply:
As we allow for implementations to support option values beyond the ones we explicitly define, the easiest way of making this optional is to leave it out for now. This would also more clearly allow for an implementation that did not support "hide if whole" fraction display to emit an error, rather than be required to fail silently.

> This value is supported by ICU.

Could you give me a pointer to how/where this is supported? I could not find this myself.

Reply:
Ah, thank you, I was looking for it among the precision options, as that's how it's presented here. Finding it for ICU let me also find the corresponding [JS option](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat/NumberFormat#trailingzerodisplay), which to no-one's surprise has a different spelling: `trailingZeroDisplay: 'stripIfInteger'`.

So let me amend my position here, and reiterate what I've stated previously: I'm fine with any subset of the Intl formatters and their options. In this case, I think we should add this as a separate option, following the TC39 nomenclature. This will also allow us to add it for `:number`.

I would be happy for us to include other rounding options for `:number` and related functions, which were previously left out.

Reply:
I can add the separate option using the Intl spelling. Do you have a list of other options to add?

Reply:
We're missing the following, which could well be added in a separate PR:
- `roundingPriority`
  - `auto` (default)
  - `morePrecision`
  - `lessPrecision`
- `roundingIncrement`
  - 1 (default), 2, 5, 10, 20, 25, 50, 100, 200, 250, 500, 1000, 2000, 2500, and 5000
- `roundingMode`
  - `ceil`
  - `floor`
  - `expand`
  - `trunc`
  - `halfCeil`
  - `halfFloor`
  - `halfExpand` (default)
  - `halfTrunc`
  - `halfEven`

Reply:
I can't do this right this second, but will add these to this PR. Should they be optional or standard? I'm leaning towards standard. 

Almost all of our work between now and 46.1 appears to be fixing up the registry/function set, plus some housekeeping.

Reply:
> Should they be optional or standard?

I'm fine either way.

> Almost all of our work between now and 46.1 appears to be fixing up the registry/function set, plus some housekeeping.

Good!

Reply:
Added. I made them standard.

@mihnita @catamorphism Please note this (proposed) change.